File: ruby/lib/number_parser.rb
require 'digit_parser'

class NumberParser
  
  def initialize(input)
    @input = input
  end

  def value
    @value ||= parse
  end
  
  def illegible?
    value.include? '?'
  end
  
  def valid?
    !illegible? && checksum == 0
  end
  
  private
  
    def checksum
      digit_array = value.chars.map(&:to_i)
      digit_array.each_with_index.inject(0) { |sum, (digit,i)| sum + digit * (9-i) } % 11
    end
  
    def parse
      (0..8).inject('') { |result, i| result += DigitParser.new.parse(digit_at(i)) }
    end
  
    def digit_at(i)
      @input[i*3..2+i*3] + @input[27+i*3..29+i*3] + @input[54+i*3..56+i*3]
    end
end

File: javascript/src/number_parser.js
function NumberParser(input) {
  
  var _input = input
  var _digit_parser = new DigitParser

  var value = function() {
    var result = ''
    for(i=0; i<=8; i++) {
      result += _digit_parser.parse(digit_at(i))
    }
    return(result);
  }
  
  var is_illegible = function() {
    return(value().includes('?'));
  }

  var is_valid = function() {
    return(!is_illegible() && checksum() == 0);
  }

  // private

  var checksum = function() {
    var digit_array = value().split('').map(function(val) { return(parseInt(val)); });
    return(digit_array.reduce(function(sum, digit, i, arr) { return(sum + digit * (9-i)) }, 0) % 11);
  }

  var digit_at = function(i) {
    return(_input.slice(i*3, 3+i*3) + _input.slice(27+i*3, 30+i*3) + _input.slice(54+i*3, 57+i*3));
  }

  // Public API
  this.value = value;
  this.is_illegible = is_illegible;
  this.is_valid = is_valid;
}